news/trn/trn.1
.\" $Id: trn.1,v 1.1 1993/07/19 20:07:08 nate Exp $
.\" 
.\" This software is Copyright 1991 by Stan Barber. 
.\"
.\" Permission is hereby granted to copy, reproduce, redistribute or otherwise
.\" use this software as long as: there is no monetary profit gained
.\" specifically from the use or reproduction of this software, it is not
.\" sold, rented, traded or otherwise marketed, and this copyright notice is
.\" included prominently in any copy made. 
.\"
.\" The author make no claims as to the fitness or correctness of this software
.\" for any use whatsoever, and it is provided as is. Any use of this software
.\" is at the user's own risk. 
.\" 
.de Sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.\" unbreakable dash.
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
'br\}
.TH TRN 1 LOCAL
.UC 6
.SH NAME
trn - threaded read news program
.SH SYNOPSIS
.B trn [options] [newsgroups]
.SH DESCRIPTION
.I Trn
is a threaded version of
.I rn,
which is a replacement for the readnews(1) program.
Being \*(L"threaded\*(R" means that the articles are interconnected
in reply order.
Each discussion thread is a tree of articles where all the
reply (child) articles branch off from their respective originating
(parent) articles.
A representation of this tree (or a portion of it) is displayed in the
article header as you are reading news.
This gives you a better feel for how all the articles are related, and
even lets you see at a glance when an article has replies \*(-- a good
thing to check before posting.
In addition,
.I trn
has an thread selector that allows you to quickly browse through a
list of subjects and choose the ones you find interesting.
This thread selector can be sorted by in a variety of orders and switched
into various display modes that allows you to pick all the subjects
separately (threads can have multiple subjects) or even pick individual
articles.
Any items you don't select can be saved for reading later or
marked as read with a single keystroke.
.PP
If you are already familiar with
.I trn
you may just want to read the
.IR "WHAT'S NEW" 
section.
People upgrading from
.I rn
will probably want to pay attention to the sections on
.IR "The Selector" ,
.IR "The Tree Display" ,
and the aforementioned
.IR "WHAT'S NEW" .
If you're impatient, just dive in and get started.
All the regular commands will be familiar to an
.I rn
or
.I trn
user, and the on-line help will give you a quick run-down of what commands
are available (just type \*(L'h\*(R' from any prompt).
I'd also suggest using the command:
.Sp
    trn \-x \-X
.Sp
to make sure some of the best features are turned on.
.Sh "Starting Trn"
If no newsgroups are specified, all the newsgroups which have unread news
will be presented to the user in the order in which they occur in the
.I .newsrc
file.
At the prompt for each group you can choose to read it, skip it, move it,
etc.
If a list of newsgroups is provided on the command line,
.I trn
will start up in \*(L"add\*(R" mode, using the list as a set of patterns to
add new newsgroups and restrict which newsgroups are displayed (see also
the discussion of the \*(L'a\*(R' command on the newsgroup-selection level).
.PP
.I Trn
operates on four levels: the newsgroup-selection level, the thread
selector, the article-reading level, and the paging level.
Each level has its own set of commands, and its own help menu.
At the paging level (the bottom level)
.I trn
behaves much like the
.IR more (1)
program.
At the article-reading level articles are presented to you in the
order of their replies, with the subjects being ordered by the date
of the oldest unread article (though there are commands for changing
the default display order).
In the thread selector you are presented with the subjects and
(usually) authors associated with each discussion thread, and given
a chance to choose which ones you wish to read now, save for later, or
manipulate in some way.
At the newsgroup-selection level (the top level), you may specify which
newsgroup you want next, or read them in the default order, which is the
order that the newsgroups occur in your
.I .newsrc
file.
(You will therefore want to rearrange your
.I .newsrc
file to put the most interesting newsgroups first.
This can be done with the \*(L'm\*(R' command on the Newsgroup Selection level.
WARNING: invoking readnews/vnews (the old user interface) in any way (including
as a news checker in your login sequence!) will cause your
.I .newsrc
to be disarranged again.)
.PP
On any level, at ANY prompt, help is available by typing an \*(L'h\*(R'.
This gives you a summary of available commands and what they do.
Remember this command, you'll need it.
.PP
Typing space to any question means to do the normal thing.
You will know what that is because
every prompt has a list of several plausible commands enclosed in
square brackets.
The first command in the list is the one which will be done if you type
a space.
(All input is done in cbreak mode, so carriage returns should not be typed
to terminate anything except certain multi-character commands.
Those commands will be obvious in the discussion below because they take an
argument.)
.PP
Upon startup,
.I trn
will do several things:
.Ip 1. 4
It will look for your
.I .newsrc
file, which is your list of subscribed-to
newsgroups.
If
.I trn
doesn't find a
.IR .newsrc ,
it will create one.
If it does find one, it will back it up under the name \*(L".\|oldnewsrc\*(R".
.Ip 2. 4
It will input your
.I .newsrc
file, listing out the first several newsgroups
with unread news.
.Ip 3. 4
It will perform certain consistency checks on your
.IR .newsrc .
If your
.I .newsrc
is out of date in any of several ways,
.I trn
will warn you and patch it up for you, but you may have to wait a little
longer for it to start up.
.Ip 4. 4
.I Trn
will next check to see if any new newsgroups have been created, and give
you the opportunity to add them to your
.IR .newsrc .
.Ip 5. 4
.I Trn
goes into the top prompt level \*(-- the newsgroup-selection level.
.Sh "Newsgroup Selection Level"
In this section the words \*(L"next\*(R" and \*(L"previous\*(R" refer to
the ordering of the newsgroups in your
.I .newsrc
file.
On the newsgroup-selection level, the prompt looks like this:
.Sp
====== 17 unread articles in talk.blurfl \*(-- read now? [ynq]
.Sp
unless the group is set for unthreaded reading, in which case the first
six characters are \*(L"******\*(R".
The following commands may be given at this level:
.Ip \+ 8
Enter this newsgroup through the selector.
.Ip y 8
Begin reading this newsgroup now.
.Ip SP 8
Enter the newsgroup by executing the default command listed in []'s.
.Ip .command 8
Do this newsgroup now, but execute
.I command
before displaying anything.
The command will be interpreted as if typed on the article selection level.
.Ip = 8
Start this newsgroup, but list subjects before displaying articles.
.Ip U 8
Enter this newsgroup through the \*(L"Set unread\*(R" prompt.
.Ip t 8
Toggle the newsgroup between threaded and unthreaded reading.
The default is threaded, and the current setting is stored in your .newsrc.
.Ip n 8
Go to the next newsgroup with unread news.
.Ip N 8
Go to the next newsgroup.
.Ip p 8
Go to the previous newsgroup with unread news.
If there is none, stay at the current newsgroup.
.Ip P 8
Go to the previous newsgroup.
.Ip \- 8
Go to the previously displayed newsgroup (regardless of whether it is
before or after the current one in the list).
.Ip 1 8
Go to the first newsgroup.
.Ip ^ 8
Go to the first newsgroup with unread news.
.Ip $ 8
Go to the end of the newsgroups list.
.Ip "g newsgroup" 8
Go to
.IR newsgroup ,
which can be the group's name or a zero-relative number of the groups in
your .newsrc (see the \*(L'L\*(R' command to list your .newsrc).
If it isn't currently subscribed to, you will be asked if you want to
subscribe.
.Ip "/pattern" 8
Scan forward for a newsgroup matching
.IR pattern .
Patterns do globbing like filenames, i.\|e., use ? to match a single
character, * to match any sequence of characters, and [] to specify a list
of characters to match.
(\*(L"all\*(R" may be used as a synonym for \*(L"*\*(R".)
Unlike normal filename globbing, newsgroup-searching is not anchored to
the front and back of the filename, i.\|e. \*(L"/ski\*(R" will find
rec.skiing.
You may use ^ or $ to anchor the front or back of the search:
\*(L"/^test$\*(R" will find newsgroup test and nothing else
If you want to include newsgroups with 0 unread articles, append /r.
If the newsgroup is not found between the current newsgroup and the last
newsgroup, the search will wrap around to the beginning.
.Ip "?pattern" 8
Same as /, but search backwards.
.Ip u 8
Unsubscribe from the current newsgroup.
.Ip "l string" 8
List newsgroups not subscribed to which contain the string specified.
.Ip L 8 13v
Lists the current state of the
.IR .newsrc ,
along with status information.
.Sp
.nf
    \h'|0.5i'Status    \h'|2i'Meaning
    \h'|0.5i'<number>  \h'|2i'Count of unread articles in newsgroup.
    \h'|0.5i'READ      \h'|2i'No unread articles in newsgroup.
    \h'|0.5i'UNSUB     \h'|2i'Unsubscribed newsgroup.
    \h'|0.5i'BOGUS     \h'|2i'Bogus newsgroup.
    \h'|0.5i'JUNK      \h'|2i'Ignored line in .newsrc
\h'|2i'(e.\|g. readnews \*(L"options\*(R" line).
.fi
.Sp
(A bogus newsgroup is one that is not in the list of active newsgroups
in the active file, which on most systems is /usr/lib/news/active unless
you use NNTP.)
.Ip "m {name}" 8
Move the named newsgroup somewhere else in the
.IR .newsrc .
If no name is given, the current newsgroup is moved.
There are a number of ways to specify where you want the newsgroup \*(-- type
h for help when it asks where you want to put it.
.Ip c 8
Catch up \*(-- mark all unread articles in this newsgroup as read.
.Ip A 8
Abandon the changes made to the current newsgroup since
.I trn
was started.
Useful when you accidentally mark a group as read.
.Ip "o {pattern}" 8
Only display those newsgroups whose name matches
.IR pattern .
Patterns are the same as for the \*(L'/\*(R' command.
Multiple patterns may be separated by spaces, just as on the
command line.
The restriction will remain in effect either until there are no articles
left in the restricted set of newsgroups, or another restriction command
is given.
Since
.I pattern
is optional, \*(L'o\*(R' by itself will remove the
restriction.
.Ip "a pattern" 8
Add unsubscribed newsgroups matching
.IR pattern .
If any matching newsgroups are found, you will be asked for each one whether
you would like to add it.
If you want to add all the newsgroups, you can
type \*(L'Y\*(R' and they will be added the the end of the
.I .newsrc
file.
If you don't want to subscribe, all
the remaining groups can be ignored by typing \*(L'N\*(R'.
After any new newsgroups have been added, the \*(L'a\*(R' command also
restricts the current set of newsgroups just like the \*(L'o\*(R' command
does.
.Ip & 8
Print out the current status of command-line switches and any newsgroup
restrictions.
.Ip "&switch {switch}" 8
Set additional command-line switches.
.Ip && 8
Print out the current macro definitions.
.Ip "&&keys commands" 8
Define additional macros.
.Ip !command 8
Escape to a subshell.
One exclamation mark (!) leaves you in your own news directory.
A double exclamation mark (!!) leaves you in the spool
directory for news, which is usually /usr/spool/news unless you're
using NNTP to read news.
The environment variable SHELL will be used if defined.
If
.I command
is null, an interactive shell is started.
.Ip v 8
Print the current version number and information on where to send bug
reports.
.Ip q 8
Quit.
.Ip x 8
Quit, restoring .newsrc to its state at startup of
.IR trn .
The .newsrc you would have had if you had exited with \*(L'q\*(R' will be
called .newnewsrc, in case you didn't really want to type \*(L'x\*(R'.
.Ip ^K 8
Edit the global list of memorized commands (in the global KILL file) that
you wish to be performed in every newsgroup as it is started up (that is, when
it is selected at the newsgroup-selection level).
This file contains commands (one per line) such as /subject/:j or /author/f:+
to kill or select articles based on the indicated search criteria.
There is also a local list of commands for each newsgroup that can contain
kill/selection commands tailored for each specific group.
Because of the overhead involved in searching for articles to kill, it is
better if possible to use a local list rather than the global one.
Local memorized commands are usually maintained by using the \*(L'A\*(R'
or \*(L'T\*(R' commands from the article/pager level or in the selector.
There is also a K search modifier that appends any search command you
desire to add.
It is also possible to manually edit the file with the \*(L'^K\*(R'
command from anywhere inside a newsgroup.
If either of the environment variables VISUAL or EDITOR is set, the
specified editor will be invoked; otherwise a default editor
is invoked on the KILL file.
.Sh "The Selector"
Most people who don't have all day to read news will want to enter a newsgroup
by way of the selector.
This is accomplished by using the \*(L'+\*(R'
command at the newsgroup-selection or article/pager levels.
In fact, this may be the default command for entering a newsgroup, depending
on how your version of
.I trn
was configured and your use of the
.B \-X
option.
.PP
The selector displays a list of articles by their subjects and (usually)
authors.
The articles are grouped into threads by default (which may list multiple
subjects per selectable item if the subject has changed during the discussion)
and ordered by the date of their oldest unread article.
Thread or subject groups are also shown with a count of the number of
articles in each group.
Each selectable item is preceded by a letter or number that can be typed
to toggle its selection.
Items that are selected are flagged with a \*(L'+\*(R' after their letter.
Groups that have only some of their articles selected are flagged with
a \*(L'*\*(R'.
You can change the selector's mode (to pick each subject separately or
pick individual articles), order the list by a variety of sort
criteria, and switch the author display between its long, medium and short
styles using the commands detailed below.
.PP
The following commands are available in the selector:
.Ip "a-z,0-9,A-Z" 8
Select/deselect the indicated item by its letter or number.
There are quite a few letters omitted from the alpha characters to be
typed as commands \*(-- see below.
Also, the variable SELECTCHARS is available to customize which characters
you want to be used as selection letters, overriding their command function.
.Ip SP 8
Perform the default command.
This is usually > for most pages, and Z on the last page (although D and X
are also quite popular).
.Ip CR 8
Begin reading.
If no articles are selected, the current item is selected (unless you've
marked it as killed).
.Ip "Z,TAB" 8
Begin reading.
If no articles are selected, read all unread articles.
.Ip "\*(L'.\*(R'" 8
Toggle the current item's selection (the one under the cursor).
.Ip * 8
Same as \*(L'.\*(R' except that it affects all articles with the same
subject (useful in the article selector).
.Ip # 8
Make an overriding selection that reads the current item only,
temporarily ignoring all other selections.
.Ip "k, \*(L',\*(R'" 8
Mark the current item as killed.
.Ip "m, \e" 8
Unmark the current item.
.Ip \- 8
Set a range, as in a \- k.
Repeats the last marking action: selection, deselection, killing,
or unmarking.
.Ip @ 8
Toggle all visible selections.
.Ip M 8
Mark the current item's article(s) to return on newsgroup exit
and kill the item.
.Ip Y 8
Yank back and select the marked-to-return articles, clearing their
to-return status.
.Ip E 8
Exclude all unselected items from the selection list (narrow the display).
Press it again to pick from all available items.
.Ip "n, ]" 8
Move down to the next item (try the down-arrow keypad key also).
.Ip "p, [" 8
Move up to the previous item (try the up-arrow keypad key also).
.Ip < 8
Go to previous page (try the left-arrow keypad key also).
.Ip > 8
Go to next page (try the right-arrow keypad key also).
.Ip ^ 8
Go to the first page.
.Ip $ 8
Go to the last page.
.Ip S 8
Set the items the selector displays: threads, subjects or articles.
If the group is unthreaded setting this to threads will thread the
group.
.Ip = 8
Switch between the article selector and the subject/thread selector.
.Ip O 8
Pick the order for the items: date, subject, author, item count
(for thread/subject groups), and a subject-date grouping of
individual articles.
Typing the selection in lower-case will sort the articles in the
default direction, while using upper-case will reverse the sort.
There is a separate default sort order for the subject/thread selector
and the article selector.
See the
.B \-O

news/trn/trn.c
/*  trn -- threaded readnews program based on rn 4.4
 *
 *  Author/Maintainer of trn: davison@borland.com (Wayne Davison)
 *  Organization: Borland International
 *  Author/Maintainer of rn: sob@bcm.tmc.edu (Stan Barber)
 *  Organization: Baylor College of Medicine, Houston,Tx
 *  Original Author: lwall@sdcrdcf.UUCP (Larry Wall)
 *  Organization: System Development Corporation, Santa Monica
 *
 *  History:
 *	01/14/83 - rn begun
 *	04/08/83 - rn 1.0
 *	09/01/83 - rn 2.0
 *	05/01/85 - rn 4.3
 *	11/01/89 - rn/rrn integration
 *	11/25/89 - trn begun
 *	07/21/90 - trn 1.0
 *	07/04/91 - rn 4.4
 *	11/25/91 - trn 2.0
 *	03/01/93 - trn 3.0 (or so... :-) )
 */
/* This software is Copyright 1991 by Stan Barber. 
 *
 * Permission is hereby granted to copy, reproduce, redistribute or otherwise
 * use this software as long as: there is no monetary profit gained
 * specifically from the use or reproduction of this software, it is not
 * sold, rented, traded or otherwise marketed, and this copyright notice is
 * included prominently in any copy made. 
 *
 * The authors make no claims as to the fitness or correctness of this software
 * for any use whatsoever, and it is provided as is. Any use of this software
 * is at the user's own risk. 
 */

#include "patchlevel.h"
static char rnid[] = "@(#)$Id: trn.c,v 1.3 1993/11/17 23:04:10 nate Exp $";
static char patchlevel[] = PATCHLEVEL;

#include "INTERN.h"
#include "common.h"
#include "trn.h"
#include "EXTERN.h"
#include "rcstuff.h"
#include "term.h"
#include "final.h"
#include "search.h"
#include "ngdata.h"
#include "ngstuff.h"
#include "util.h"
#include "only.h"
#include "ngsrch.h"
#include "help.h"
#include "last.h"
#include "init.h"
#include "intrp.h"
#include "rcln.h"
#include "sw.h"
#include "cache.h"
#include "addng.h"
#include "ng.h"
#include "kfile.h"
#include "nntp.h"

void
rn_init()
{
    ;
}

void
main(argc,argv)
int argc;
char *argv[];
{
    bool foundany;
    register char *s;
    bool oh_for_the_good_old_days = FALSE;
    int direction = 1;

#if !THREAD_INIT
    /* Default to threaded operation if our name starts with a 't' */
    if ((s = rindex(argv[0],'/')) == Nullch)
	s = argv[0];
    else
	s++;
    if (*s == 't')
	use_threads = TRUE;
    else
	select_on = 0;
#endif
    foundany = initialize(argc,argv);

    if (maxngtodo)
	starthere = 0;
    else if (!foundany) {		/* nothing to do? */
#ifdef VERBOSE
	if (verbose)
	    fputs("\
No unread news in subscribed-to newsgroups.  To subscribe to a new\n\
newsgroup use the g<newsgroup> command.\n\
",stdout) FLUSH;
#endif
	starthere = nextrcline;
    }

    /* loop through all unread news */

    {
	bool special = FALSE;		/* temporarily allow newsgroup */
					/*   with no unread news? */
	bool retry;			/* cycle back to top of list? */
	NG_NUM recent_ng = 0;
	
	current_ng = 0;
	do {
	    retry = FALSE;
	    if (findlast) {
		findlast = FALSE;
		starthere = 0;
		if (*lastngname) {
		    if ((ng = find_ng(lastngname)) == nextrcline)
			ng = 0;
		    else {
			set_ngname(lastngname);
		    	set_toread(ng);
			if (toread[ng] <= TR_NONE)
			    ng = 0;
		    }
		}
	    } else {
		ng = starthere;
		starthere = 0;
	    }
	    while (ng <= nextrcline) {	/* for each newsgroup */
		if (ng == nextrcline) {	/* after the last newsgroup? */
		    mode = 'f';
#ifdef ONLY
		    if (maxngtodo) {
			if (retry)
#ifdef VERBOSE
			    IF(verbose)
				printf("\nRestriction %s%s still in effect.\n",
				    ngtodo[0],
				    maxngtodo > 1 ? ", etc." : nullstr) FLUSH;
			    ELSE
#endif
#ifdef TERSE
				fputs("\n(\"Only\" mode.)\n",stdout) FLUSH;
#endif
			else {
#ifdef VERBOSE
			    IF(verbose)
				fputs("\nNo articles under restriction.",
				  stdout) FLUSH;
			    ELSE
#endif
#ifdef TERSE
				fputs("\nNo \"only\" articles.",stdout) FLUSH;
#endif
			    end_only();	/* release the restriction */
			    retry = TRUE;
			}
		    }
#endif
		} else {
		    bool shoe_fits;	/* newsgroup matches restriction? */

		    mode = 'n';
		    if (toread[ng] >= TR_NONE) {	/* recalc toread? */
			set_ngname(rcline[ng]);
			shoe_fits = inlist(ngname);
			if (shoe_fits)
			    set_toread(ng);
			if (paranoid) {
			    recent_ng = current_ng;
			    current_ng = ng;
			    cleanup_rc();
					/* this may move newsgroups around */
			    ng = current_ng;
			    set_ngname(rcline[ng]);
			}
		    }
		    if (toread[ng] < (emptyOnly || special ? TR_NONE : TR_ONE)
		     || !shoe_fits) {		/* unwanted newsgroup? */
			ng += direction;	/* then skip it */
			if (ng < 0) {
			   ng = 1;
			   direction = 1;
			}
			continue;
		    }
		}
		special = FALSE;	/* go back to normal mode */
		if (ng != current_ng) {
		    recent_ng = current_ng;
					/* remember previous newsgroup */
		    current_ng = ng;	/* remember current newsgroup */
		}
    reask_newsgroup:
		unflush_output();	/* disable any ^O in effect */
		if (ng >= nextrcline) {
#ifdef USE_NNTP
		    if (actFetchTime
		     && time(Null(time_t*)) - lastactfetch > actFetchTime) {
			fclose(actfp);
			ngdata_init();	/* re-grab the active file */
		    }
#endif
		    dfltcmd = (retry ? "npq" : "qnp");
#ifdef VERBOSE
		    IF(verbose)
			printf("\n****** End of newsgroups -- what next? [%s] ",
			    dfltcmd);
		    ELSE
#endif
#ifdef TERSE
			printf("\n**** End -- next? [%s] ", dfltcmd);
#endif
		} else {
		    ThreadedGroup = (use_threads && rcchar[ng] != '0');
		    dfltcmd = (select_on
			&& (ART_NUM)toread[ng] >= select_on ? "+ynq" : "ynq");
#ifdef VERBOSE
		    IF(verbose)
			printf("\n%s %3ld unread article%s in %s -- read now? [%s] ",
			    ThreadedGroup? "======" : "******",
			    (long)toread[ng], (toread[ng]==TR_ONE ? nullstr : "s"),
			    ngname, dfltcmd);
		    ELSE
#endif
#ifdef TERSE
			printf("\n%s %3ld in %s -- read? [%s] ",
			    ThreadedGroup? "====" : "****",
			    (long)toread[ng],ngname,dfltcmd);
#endif
		}
		fflush(stdout);
    reinp_newsgroup:
		eat_typeahead();
		getcmd(buf);
		if (errno || *buf == '\f') {
		    putchar('\n') FLUSH; /* if return from stop signal */
		    goto reask_newsgroup;	/* give them a prompt again */
		}
		setdef(buf,dfltcmd);
#ifdef VERIFY
		printcmd();
#endif
    do_command:
		direction = 1;		/* default to forward motion */
		addnewbydefault = 0;
		switch (*buf) {
		case 'P':		/* goto previous newsgroup */
		    special = TRUE;	/* don't skip it if toread==0 */
		    /* FALL THROUGH */
		case 'p':		/* find previous unread newsgroup */
		    if (ng > 0)
			ng--;
		    direction = -1;	/* go backward in the newsrc */
		    break;
		case '-':
		    ng = recent_ng;	/* recall previous newsgroup */
		    if (ng < nextrcline)
			if (!get_ng(rcline[ng],0))
			    ng = current_ng;
		    special = TRUE;	/* don't skip it if toread==0 */
		    break;
		case 'q': case 'Q': case 'x':	/* quit? */
		    oh_for_the_good_old_days = (*buf == 'x');
		    putchar('\n') FLUSH;
		    ng = nextrcline+1;	/* satisfy */
		    retry = FALSE;	/*   loop conditions */
		    break;
		case '^':
		    putchar('\n') FLUSH;
		    ng = 0;
		    break;
		case 'n':		/* find next unread newsgroup */
		    if (ng == nextrcline) {
			putchar('\n') FLUSH;
			retry = TRUE;
		    } else if (toread[ng] > TR_NONE)
			retry = TRUE;
		    ng++;
		    break;
		case 'N':		/* goto next newsgroup */
		    special = TRUE;	/* and don't skip it if toread==0 */
		    ng++;
		    break;
		case '1':		/* goto 1st newsgroup */
		    ng = 0;
		    special = TRUE;	/* and don't skip it if toread==0 */
		    break;
		case '$':
		    ng = nextrcline;	/* goto last newsgroup */
		    retry = TRUE;
		    break;
		case 'L':
		    list_newsgroups();
		    goto reask_newsgroup;
		case '/': case '?':	/* scan for newsgroup pattern */
#ifdef NGSEARCH
		    switch (ng_search(buf,TRUE)) {
		    case NGS_ERROR:
			goto reask_newsgroup;
		    case NGS_ABORT:
			goto reinp_newsgroup;
		    case NGS_INTR:
#ifdef VERBOSE
			IF(verbose)
			    fputs("\n(Interrupted)\n",stdout) FLUSH;
			ELSE
#endif
#ifdef TERSE
			    fputs("\n(Intr)\n",stdout) FLUSH;
#endif
			ng = current_ng;
			goto reask_newsgroup;
		    case NGS_FOUND:
			special = TRUE;	/* don't skip it if toread==0 */
			break;
		    case NGS_NOTFOUND:
#ifdef VERBOSE
			IF(verbose)
			    fputs("\n\nNot found -- use a or g to add newsgroups\n",
				stdout) FLUSH;
			ELSE
#endif
#ifdef TERSE
			    fputs("\n\nNot found\n",stdout) FLUSH;
#endif
			goto reask_newsgroup;
		    }
#else
		    notincl("/");
#endif
		    break;
		case 'm':
#ifndef RELOCATE
		    notincl("m");
		    break;
#endif		    
		case 'g':	/* goto named newsgroup */
		    if (!finish_command(FALSE))
					/* if they didn't finish command */
			goto reinp_newsgroup;	/* go try something else */
		    for (s = buf+1; *s == ' '; s++);
					/* skip leading spaces */
#ifdef RELOCATE
		    if (!*s && *buf == 'm' && ngname && ng < nextrcline)
			strcpy(s,ngname);
#endif
		    {
		        char *_s;
			for (_s=s; isdigit(*_s); _s++)
			    ;
			if (isspace(*_s))
			    *_s = '\0';
		        if (*_s)
			    /* found non-digit before hitting end */
			    set_ngname(s);
			else {
			    int rcnum;
			    rcnum = atoi(s);
			    if (rcnum < nextrcline)
				set_ngname(rcline[rcnum]);
			    else {
				printf("\nOnly %d groups. Try again.\n",
					nextrcline) FLUSH;
				goto reask_newsgroup;
			    }
			}
		    }
		    /* try to find newsgroup */
#ifdef RELOCATE
		    if (!get_ng(ngname,(*buf=='m'?GNG_RELOC:0) | GNG_FUZZY))
#else
		    if (!get_ng(ngname,GNG_FUZZY))
#endif
			ng = current_ng;/* if not found, go nowhere */
		    special = TRUE;	/* don't skip it if toread==0 */
		    break;
#ifdef DEBUG
		case 'D':
		    printf("\nTries: %d Hits: %d\n",
			softtries,softtries-softmisses) FLUSH;
		    goto reask_newsgroup;
#endif
		case '!':		/* shell escape */
		    if (escapade())	 /* do command */
			goto reinp_newsgroup;
					/* if rubbed out, re input */
		    goto reask_newsgroup;
		case Ctl('k'):		/* edit global KILL file */
		    edit_kfile();
		    goto reask_newsgroup;
		case 'c':		/* catch up */
#ifdef CATCHUP
		    if (ng < nextrcline)
			ask_catchup();
		    ng++;
#else
		    notincl("c");
#endif
		    break;
		case 't':
		    if (!use_threads)
			printf("\n\nNot running in thread mode.\n");
		    else if (ng < nextrcline && toread[ng] >= TR_NONE) {
			bool read_unthreaded = (rcchar[ng] == ':');
			rcchar[ng] = (read_unthreaded ? '0' : ':');
			printf("\n\n%s will be read %sthreaded.\n",
			    rcline[ng], read_unthreaded? "un" : "") FLUSH;
			set_toread(ng);
		    }
		    special = TRUE;	/* don't skip it if toread==0 */
		    break;
		case 'u':		/* unsubscribe */
		    if (ng < nextrcline && toread[ng] >= TR_NONE) {
					/* unsubscribable? */
			printf(unsubto,rcline[ng]) FLUSH;
			rcchar[ng] = NEGCHAR;
					/* unsubscribe to (from?) it */
			toread[ng] = TR_UNSUB;
					/* and make line invisible */
			ng++;		/* do an automatic 'n' */
		    }
		    break;
		case 'h': {		/* help */
		    int cmd;

		    if ((cmd = help_ng()) > 0)
			pushchar(cmd);
		    goto reask_newsgroup;
		}
		case 'A':
		    if (ng >= nextrcline)
			break;
reask_abandon:
#ifdef VERBOSE
		    IF(verbose)
			in_char("\nAbandon changes to current newsgroup? [yn] ", 'B');
		    ELSE
#endif
#ifdef TERSE
			in_char("\nAbandon? [ynh] ", 'B');
#endif
		    setdef(buf,"y");
#ifdef VERIFY
		    printcmd();
#endif
		    putchar('\n') FLUSH;
		    if (*buf == 'h') {
#ifdef VERBOSE
			printf("Type y or SP to abandon the changes to this group since you started trn.\n");
			printf("Type n to leave the group as it is.\n");
#else
			printf("y or SP to abandon changes to this group.\n");
			printf("n to forget it.\n");
#endif
			goto reask_abandon;
		    } else if (*buf != 'y' && *buf != 'n' && *buf != 'q') {
			printf(hforhelp);
			settle_down();
			goto reask_abandon;
		    } else if (*buf == 'y')
			abandon_ng(ng);
		    special = TRUE;	/* don't skip it if toread==0 */
		    break;
		case 'a':
#ifndef FINDNEWNG
		    notincl("a");
		    goto reask_newsgroup;
#else
		    /* FALL THROUGH */
#endif
		case 'o':
		case 'O':
#ifdef ONLY
		{
#ifdef FINDNEWNG
		    bool doscan = (*buf == 'a');
#endif

		    if (!finish_command(TRUE)) /* get rest of command */
			goto reinp_newsgroup;	/* if rubbed out, try something else */
		    end_only();
		    if (buf[1]) {
			bool minusd = instr(buf+1,"-d", TRUE) != Nullch;

			sw_list(buf+1);
			if (minusd)
			    cwd_check();
			putchar('\n') FLUSH;
#ifdef FINDNEWNG
			if (doscan && maxngtodo)
			    scanactive();
#endif
			emptyOnly = (*buf == 'o' && maxngtodo);
		    }
		    ng = 0;		/* simulate ^ */
		    retry = FALSE;
		    break;
		}
#else
		    notincl("o");
		    goto reask_newsgroup;
#endif
		case '&':
		    if (switcheroo()) /* get rest of command */
			goto reinp_newsgroup;	/* if rubbed out, try something else */
		    goto reask_newsgroup;
		case 'l': {		/* list other newsgroups */
		    if (!finish_command(TRUE)) /* get rest of command */
			goto reinp_newsgroup;	/* if rubbed out, try something else */
		    for (s = buf+1; *s == ' '; s++);
		    			/* skip leading spaces */
		    sprintf(cmd_buf,"%s '%s'",filexp(NEWSGROUPS),s);
		    resetty();
		    if (doshell(sh,cmd_buf))
#ifdef VERBOSE
			IF(verbose)
			    fputs("    (Error from newsgroups program)\n",
				stdout) FLUSH;
			ELSE
#endif
#ifdef TERSE
			    fputs("(Error)\n",stdout) FLUSH;
#endif
		    noecho();
		    crmode();
		    goto reask_newsgroup;
		}
		case 'U': case '+':
		case '.': case '=':
		case 'y': case 'Y': case '\t': /* do normal thing */
		    if (ng >= nextrcline) {
			fputs("\nNot on a newsgroup.",stdout) FLUSH;
			goto reask_newsgroup;
		    } else if (*buf == '.') {	/* start command? */
			if (!finish_command(FALSE)) /* get rest of command */
			    goto reinp_newsgroup;
			s = savestr(buf+1);  /* do_newsgroup will free it */
		    } else {
			s = buf;
			if (*buf == '+' || *buf == 'U' || *buf == '=')
			    *s++ = lastchar; /* restore 0200 if from a macro */
			save_typeahead(s, sizeof buf - 1);
			if (*buf)
			    s = savestr(buf);
			else
			    s = Nullch;
		    }
		    if (toread[ng])
			retry = TRUE;
		    switch (do_newsgroup(s)) {
		    case NG_ERROR:
		    case NG_NORM:
			ng++;
			break;
		    case NG_ASK:
			goto reask_newsgroup;
		    case NG_SELPRIOR:
			*buf = 'p';
			goto do_command;
		    case NG_SELNEXT:
			*buf = 'n';
			goto do_command;
		    case NG_MINUS:
			ng = recent_ng;	/* recall previous newsgroup */
			special = TRUE;	/* don't skip it if toread==0 */
			break;
		    }
		    break;
#ifdef STRICTCR
		case '\n':
		    fputs(badcr,stdout) FLUSH;
		    goto reask_newsgroup;
#endif
		case 'v':
		    printf("\n%s",rnid);
		    printf("\n%s",patchlevel);
		    printf("\nSend bugs to davison@borland.com\n") FLUSH;
		    goto reask_newsgroup;
		default:
		    printf("\n%s",hforhelp) FLUSH;
		    settle_down();
		    goto reask_newsgroup;
		}
	    }
	} while (retry);
    }

    /* now write .newsrc back out */

    write_rc();

    if (oh_for_the_good_old_days)
	get_old_rc();

    finalize(0);			/* and exit */
}

/* set current newsgroup */

void
set_ngname(what)
char *what;
{
    int len = strlen(what)+1;

    growstr(&ngname,&ngnlen,len);
    strcpy(ngname,what);
    growstr(&ngdir,&ngdlen,len);
    strcpy(ngdir,getngdir(ngname));
}

static char *myngdir;
static int ngdirlen = 0;

char *
getngdir(ngnam)
char *ngnam;
{
    register char *s;

    growstr(&myngdir,&ngdirlen,strlen(ngnam)+1);
    strcpy(myngdir,ngnam);
    for (s = myngdir; *s; s++)
	if (*s == '.')
	    *s = '/';
    return myngdir;
}
